sotodlib/io/imprinter.py

```diff
@@ -682,9 +686,132 @@ def delete_book_files(self, book):
             self.logger.warning(f"Failed to remove {book.path}: {e}")
             self.logger.error(traceback.format_exc())
 
+    def get_book_times(self, book):
+        """
+        Given a book of co-sampled data with a timing system, this function will
+        return a gap-filled list of timestamps that should be contained in the book.
+
+        Parameters
+        ----------
+        book: Book object
+        """
+        filedb = self.get_files_for_book(book)
+        t0 = book.start.timestamp()
+        t1 = book.stop.timestamp()
+        for i, files in enumerate(filedb.values()):
+            _files = sorted(files)
+            if i == 0:  # Get full list of timestamps for first file list
+                ts = []
+                for frame in itertools.chain(*[core.G3File(f) for f in _files]):
+                    if frame.type != core.G3FrameType.Scan:
+                        continue
+                    ts.append(get_timestamps(frame))
+                ts = fill_gaps(np.hstack(ts))
+
+                _t0, _t1 = ts[[0, -1]]
+            else:
+                _t0, _t1 = get_start_and_end(_files)
+            
+            t0 = max(t0, _t0)
+            t1 = min(t1, _t1)
+
+        m = (t0 <= ts) & (ts <= t1)
+        return ts[m]
+
+
 #####################
 # Utility functions #
 #####################
+def get_timestamps(frame):
+    """
+    Returns an array of timestamps for data in a G3 Frame
+
+    Parameters
+    -------------
+    frame : G3Frame
+        Scan frame containing detector data
+    
+    Returns
+    -------------
+    timestamps : np.ndarray
+        Array of timestamps (seconds) for samples in the frame
+    """
+    c0, c2 = frame['primary'].data[[3, 5]]
+    s, ns = split_ts_bits(c2)
+    # Add 20 years in seconds (accounting for leap years) to handle
+    # offset between EPICS time referenced to 1990 relative to UNIX time.
+    c2 = s + ns*1e-9 + 5*(4*365 + 1)*24*60*60
+    timestamps = np.round(c2 - (c0 / 480000) ) + c0 / 480000
+    return timestamps
+
+def fill_gaps(ts):
+    """
+    Fills gaps in an array of timestamps.
+
+    Parameters
+    -------------
+    ts : np.ndarray
+        List of timestamps of length `n`, potentially with gaps
+    
+    Returns
+    --------
+    new_ts : np.ndarray
+        New list of timestamps of length >= n, with gaps filled.
+    """
+    # Find indices where gaps occur and how long each gap is
+    dts = np.diff(ts)
+    dt = np.mean(dts)
+    missing = np.round(dts/dt - 1).astype(int)
```

Comment:
Would median be better here? I think a large gap (or a large number of smaller gaps) might skew the mean too much upward. On the other hand, the median would be affected by a large variance, but a timing system should keep that in check, right?

Also, probably an unlikely scenario, but if for some reason you get two samples close to each other (dts/dt < 0.5), then I think you end up with -1 missing samples, which would break the subsequent code. Maybe add a max(0, missing) or something to be safe?

Reply:
Sure, median would probably be more correct, I will update this. I hope large gaps like your describing would trigger some sort of warning somewhere.... Maybe we should do some sort of validation that tests that all arrays are indeed co-sampled before binding. That is not the purpose of this function but it might be a natural place to put it (like you mention above) since we're already extracting one set of timestamps.

For this scenario, I think we first have to figure out how we should actually handle this sort of event. If `dts/dt < 0.5` we may want to drop the sample. Like if we have `[..., 1, 2, 2.1, 3, 4]`, I think the resulting timestamps should still be [..., 1, 2, 3, 4]`

Reply:
Yeah those are definitely things to do down the line—no need to hold up this PR for them, I think. Nonetheless, the negative samples does mean I have to add a check in Bookbinder.